pkg/package-server-manager/controller.go

```diff
@@ -0,0 +1,148 @@
+// on the topology mode exposed by the current state of the Infrastructure cluster
+// singleton resource.
+func (r *PackageServerCSVReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("csv", req.NamespacedName)
+
+	log.Info("handling current request", "request", req.String())
+	defer log.Info("finished request reconciliation")
+
+	var infra configv1.Infrastructure
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: infrastructureName}, &infra); err != nil {
+		return ctrl.Result{}, err
+	}
+	highAvailabilityMode := getTopologyModeFromInfra(&infra)
+	log.Info("currently topology mode", "highly available", highAvailabilityMode)
+
+	required, err := manifests.NewPackageServerCSV(
+		manifests.WithName(r.Name),
+		manifests.WithNamespace(r.Namespace),
+		manifests.WithImage(r.Image),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	res, err := controllerutil.CreateOrUpdate(ctx, r.Client, required, func() error {
+		return reconcileCSV(r.Log, r.Image, required, highAvailabilityMode)
+	})
+	if err != nil {
+		log.Error(err, "failed to create or update the csv")
+		return ctrl.Result{}, nil
+	}
+	log.Info("reconciliation result", "res", res)
+
+	return ctrl.Result{}, nil
+}
+
+func reconcileCSV(log logr.Logger, image string, csv *olmv1alpha1.ClusterServiceVersion, highAvailabilityMode bool) error {
+	if csv.ObjectMeta.CreationTimestamp.IsZero() {
+		log.Info("attempting to create the packageserver csv")
+	}
+	if !validateCSV(log, csv) {
+		log.Info("updating invalid csv to use the default configuration")
+		tmp, err := manifests.NewPackageServerCSV(
+			manifests.WithName(csv.Name),
+			manifests.WithNamespace(csv.Namespace),
+			manifests.WithImage(image),
+		)
+		if err != nil {
+			return err
+		}
+		csv.Spec = tmp.Spec
+	}
+	if !ensureCSV(log, image, csv, highAvailabilityMode) {
+		log.V(3).Info("no further updates are necessary to the csv")
+	}
+
+	return nil
+}
+
+func (r *PackageServerCSVReconciler) infrastructureHandler(obj client.Object) []reconcile.Request {
+	log := r.Log.WithValues("infrastructure", obj.GetName())
+	if obj.GetName() != infrastructureName {
+		log.Info("not processing events for the non-cluster infrastructure resource")
+		return nil
+	}
+
+	log.Info("requeueing the packageserver deployment after encountering infrastructure event")
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Name:      r.Name,
+				Namespace: r.Namespace,
+			},
+		},
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PackageServerCSVReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	predicateNameFilter := predicate.NewPredicateFuncs(func(object client.Object) bool {
+		return object.GetName() == r.Name && object.GetNamespace() == r.Namespace
+	})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&olmv1alpha1.ClusterServiceVersion{}, builder.WithPredicates(predicateNameFilter)).
```

Comment:
predicates just filter the events the reconciler sees; the cache will still fill with all of the CSVs (including copied CSVs, etc)

instead you should set up a cache `Field` filter on the name/namespace of the CSV (similar to https://github.com/operator-framework/operator-lifecycle-manager/pull/2248/files but for the name, not a label)

that will ensure the server only sends you events for that object and the cache doesn't fill up with irrelevant CSVs